lib/model/message_list.dart

```diff
@@ -1,12 +1,202 @@
+
+  /// The parsed message contents, as a list parallel to [messages].
+  ///
+  /// The i'th element is the result of parsing the i'th element of [messages].
+  ///
+  /// This information is completely derived from [messages].
+  /// It exists as an optimization, to memoize the work of parsing.
+  final List<ZulipContent> contents = [];
+
+  /// The messages and their siblings in the UI, in order.
+  ///
+  /// This has a [MessageListMessageItem] corresponding to each element
+  /// of [messages], in order.  It may have additional items interspersed
+  /// before, between, or after the messages.
+  ///
+  /// This information is completely derived from [messages] and
+  /// the flags [haveOldest] and [fetchingOlder].
+  /// It exists as an optimization, to memoize that computation.
+  final QueueList<MessageListItem> items = QueueList();
+
+  int _findMessageWithId(int messageId) {
+    return binarySearchByKey(messages, messageId,
+      (message, messageId) => message.id.compareTo(messageId));
+  }
+
+  int _findItemWithMessageId(int messageId) {
+    return binarySearchByKey(items, messageId, _compareItemToMessageId);
+  }
+
+  static int _compareItemToMessageId(MessageListItem item, int messageId) {
+    switch (item) {
+      case MessageListHistoryStartItem():        return -1;
+      case MessageListLoadingItem():
+        switch (item.direction) {
+          case MessageListDirection.older:       return -1;
+          case MessageListDirection.newer:       return 1;
+        }
+      case MessageListMessageItem(:var message): return message.id.compareTo(messageId);
+    }
+  }
+
+  /// Update data derived from the content of the index-th message.
+  void _reparseContent(int index) {
+    final message = messages[index];
+    final content = parseContent(message.content);
+    contents[index] = content;
+
+    final itemIndex = _findItemWithMessageId(message.id);
+    assert(itemIndex > -1
+      && items[itemIndex] is MessageListMessageItem
+      && identical((items[itemIndex] as MessageListMessageItem).message, message));
+    items[itemIndex] = MessageListMessageItem(message, content);
+  }
+
+  /// Append [message] to [messages], and update derived data accordingly.
+  ///
+  /// The caller is responsible for ensuring this is an appropriate thing to do
+  /// given [narrow], our state of being caught up, and other concerns.
+  void _addMessage(Message message) {
+    assert(contents.length == messages.length);
+    messages.add(message);
+    contents.add(parseContent(message.content));
+    assert(contents.length == messages.length);
+    _processMessage(messages.length - 1);
+  }
+
+  void _insertAllMessages(int index, Iterable<Message> toInsert) {
+    // TODO parse/process messages in smaller batches, to not drop frames.
+    //   On a Pixel 5, a batch of 100 messages takes ~15-20ms in _insertAllMessages.
+    //   (Before that, ~2-5ms in jsonDecode and 0ms in fromJson,
+    //   so skip worrying about those steps.)
```

Comment:
Could this strategy become counterproductive with extremely long `messages`/`contents`? Those are just `List`s, so I expect inserting at index 0 (as in `fetchOlder`) to be linear time in the length of those lists.

Ah, though I guess this doesn't say _inserting_ in smaller batches, just parsing/processing. So maybe this means still doing one large insert, but of unparsed/unprocessed placeholders, and then processing those placeholders in batches.

Reply:
Yeah, or perhaps better yet: don't insert until we have fully-processed items to insert, but do the parsing/processing in smaller batches and then insert them all when we're done.

If processing the whole overall batch takes a frame or two, the extra wait of a frame or two isn't too bad — the network request will often/typically have more latency than that, and we compensate for the latency by just starting the request early before the user scrolls all the way to the edge. So probably best not to add the complexity of having a concept of placeholder items.

---

Or really probably the right answer at that point is to just turn `messages` and `contents` into `QueueList`s too :-)